### PR TITLE
Add translation resolver api

### PR DESCRIPTION
**Pull request by erdlet**
Adds an extensible API for resolving translations from different resources.

Unfortunately, I wasn't able to get the showcase to run on Tomcat or Payara with the current `pom.xml`, so I excluded the Tomcat dependencies and use plain JEE 8 at the moment. Maybe it makes sense to provide an example for Tomcat (for which the dependencies need a review) and an extra showcase for JEE / Jakarta EE.

### CODE REVIEW

**Comment by chkal** [comment]
Thanks a lot! I'll have a closer look ASAP. But I guess I won't find much time before the weekend.

**Comment by erdlet** [comment]
No problem ;) Looking forward to your feedback

**Comment by chkal** [comment]
Hi @erdlet,

first of all: Sorry for the delayed reply. I was quite busy for personal reasons.

Thanks a lot for creating this draft of an API for resolving I18N texts. I really think that this feature would be very valuable for MVC applications.

A few thoughts about your draft:

- I wonder if we could shorten the expression a bit which is used for resolving. Perhaps just `${toolbox.i18n('key')}`. Or even `${toolbox.t('key')}` with `t` being a shortcut for *translate*? That would be similar to [I18next](https://www.i18next.com/), which is popular in the JavaScript world.
- I guess the API should support placeholders via `MessageFormat` by default. Unfortunately EL doesn't provide support for varargs, so we would have to do something like this: `${toolbox.t('total-price', [12.99, 'EUR'])}`. A bit ugly, but maybe acceptable?

One thing I'm not completely sure about is whether having a single resolver chain for the entire application works very well for larger applications. Let me try to explain what I'm talking about. I saw a few real world applications which had all translations in a single huge properties file per language. That didn't scale very well and resulted in a huge number of unused keys after a few years. I'm not sure what the best way to deal with this is. But maybe splitting the translations into multiple chunks (maybe one per logical page) may work better. In this case having multiple 'translation container instances' may be a better way to go. 

I guess my description isn't easy to understand. Maybe some pseudo code will help:

```java
public class I18NBundleProducer {

  @Inject
  private DataSource dataSource;

  @Produces
  @Named("page1")
  @ApplicationScoped
  public TranslationResolver getBundle1() {
    // reads "bundle1.properties" from the classpath
    return new ResourceBundleResolver("bundle1");
  }

  @Produces
  @Named("page2")
  @ApplicationScoped
  public TranslationResolver getBundle2() {
    // reads "some/dir/bundle2.properties"
    return new ResourceBundleResolver("some/dir/bundle2");
  }

  @Produces
  @Named("page3")
  @ApplicationScoped
  public TranslationResolver getBundle3() {
    // reads from a table called "my_i18n_table" having two columns "key" and "value"
    return new JDBCTranslationResolver(dataSource, "my_i18n_table", "key", "value");
  }

}
```

Then, in a JSP:

```jsp
<div>
  ${page1.t('some-key')}
</div>
<div>
  ${page2.t('other-key', [12.99, 'EUR'])}
</div>
```

Following this approach, perhaps there would not be a need for something like a resolver chain. Instead, most functionality would be provided by some abstract base class, which is extended by concrete implementations which handle fetching the translation data from some source. Another good thing about having multiple resolvers would be that the keys don't need to be globally unique. So you could have a key `title` in `page1` and use the same in `page2` which may simplify the process of defining those keys.

But don't get me wrong. I don't necessarily think that this approach is superior. It is just something that I had in my mind and never found time to prototype it. I'm pretty sure both approaches have pros/cons. Maybe merging both approaches would be a good way to go?

I would love to hear your thoughts.

**Comment by erdlet** [comment]
Hi Christian,

thanks for that detailed feedback!

> I wonder if we could shorten the expression a bit which is used for resolving. Perhaps just 
> ${toolbox.i18n('key')}. Or even ${toolbox.t('key')} with t being a shortcut for translate? That would > be similar to I18next, which is popular in the JavaScript world.

Good idea, I already thought about something shorter but had no good idea. I think the ` ${toolbox.t('key')}` is good, as we don't waste space for a attribute name and frontend developers may know it from JavaScript (I wasn't aware of it but now I learned something helpful again, thanks!).

> I guess the API should support placeholders via MessageFormat by default. Unfortunately EL 
> doesn't provide support for varargs, so we would have to do something like this: 
> ${toolbox.t('total-price', [12.99, 'EUR'])}. A bit ugly, but maybe acceptable?

Absolute acceptable! I wasn't aware of `MessageFormat` and didn't want to implement such thing by myself, so I skipped to add a placeholder feature. The syntax in the template may be ugly, but I'm not aware of any template language where this looks good, so I don't see there any problem :) 

So I'd definitively change the call to `toolbox.t()` and try to add this placeholder feature (maybe as extra issue, so this one is not getting too big).

> I saw a few real world applications which had all translations in a single huge properties file per 
> language. That didn't scale very well and resulted in a huge number of unused keys after a few 
> years.

I think that's a point I missed in my design, as the last really big application where I had to work with translations was build with Wicket and that stores its translations per component. I think your approach is really good, maybe we can implement some simple default provider which is called via `${toolbox.t('key')}` or `${toolbox.t('key', ['var1', 'var2'])}` and reads from a big properties file for each language (makes it easy to implement prototypes or small applications). If someone wants to use the API for a different or advanced solution, like your example, we can provide base classes that can be used or extended.
Also, this solution will be easier to maintain as we don't have to filter and sort the beans in the chain and the user doesn't need to know how the `TranslationResolver` beans have to be scoped and prioritized to be used correct by the chain.

What do you think about this? I think we only have to remove stuff from this PR therefore, so it won't be a great effort.

**Comment by chkal** [comment]
Hi @erdlet 

> Good idea, I already thought about something shorter but had no good idea. I think the ` ${toolbox.t('key')}` is good, as we don't waste space for a attribute name and frontend developers may know it from JavaScript (I wasn't aware of it but now I learned something helpful again, thanks!).

I never used I18next myself, but I reviewed the documentation some time ago and liked it. :-)

> Absolute acceptable! I wasn't aware of `MessageFormat` and didn't want to implement such thing by myself, so I skipped to add a placeholder feature. The syntax in the template may be ugly, but I'm not aware of any template language where this looks good, so I don't see there any problem :)

AFAIK MessageFormat is quite popular for this use case. Especially, because it supports localization of numbers/dates out of the box.

> So I'd definitively change the call to toolbox.t() and try to add this placeholder feature (maybe as extra issue, so this one is not getting too big).

My hope is that integrating `MessageFormat` is just a few lines of code. However, if I'm wrong and it is more effort, splitting into a separate PR makes sense.


> I think that's a point I missed in my design, as the last really big application where I had to work with translations was build with Wicket and that stores its translations per component. I think your approach is really good, maybe we can implement some simple default provider which is called via `${toolbox.t('key')}` or `${toolbox.t('key', ['var1', 'var2'])}` and reads from a big properties file for each language (makes it easy to implement prototypes or small applications). If someone wants to use the API for a different or advanced solution, like your example, we can provide base classes that can be used or extended.

Agreed! An API similar to the one I proposed could be some kind of low-level API which provides the core functionality. But some default configuration accessible via `${toolbox.t}` is definitely fine. Reading from a single properties file from the classpath is the most common use case.

> Also, this solution will be easier to maintain as we don't have to filter and sort the beans in the chain and the user doesn't need to know how the `TranslationResolver` beans have to be scoped and prioritized to be used correct by the chain.

I agree!

PS: I just had one thought. I'm not 100% sure anymore if the sample code I provided above would work like this, as it creates multiple beans of the same type in the same scope (just with different names). But it may also be fine as long as you don't try to inject the type into some bean.

**Comment by erdlet** [comment]
I just shortened the `Toolbox#getTranslations` method to the proposed `Toolbox#t` method and also added the `MessageFormat` using resolving mechanism. Works like a charm as the demo in the showroom shows.

When I tried to remove the `TranslationResolverChain` I remembered why I've chosen this implementation strategy. As the `Locale` in `MvcContext` is resolved per request I wanted to avoid problems when injecting a `@RequestScoped` bean into an `@ApplicationScoped`. I'm not sure if this is possible in CDI without problems, because I experienced a lot of trouble doing something similar in Spring. If this is no problem in CDI, it may be easier to remove the chain.

**Comment by chkal** [comment]
> When I tried to remove the `TranslationResolverChain` I remembered why I've chosen this implementation strategy. As the `Locale` in `MvcContext` is resolved per request I wanted to avoid problems when injecting a `@RequestScoped` bean into an `@ApplicationScoped`. I'm not sure if this is possible in CDI without problems, because I experienced a lot of trouble doing something similar in Spring. If this is no problem in CDI, it may be easier to remove the chain.

Dependency injection between different scopes works fine in CDI because CDI **always** injects proxies instead of the real bean instance (only exception: `@Dependent`). So injecting a request-scoped bean into an application-scoped one works fine and without any problems. CDI has a quite different philosophy in this regard. So CDI behaves a bit like Spring with all beans being AOP-proxied.

**Comment by erdlet** [comment]
Ah good to know, thanks. Then I'll try to remove the chain and create the
APIs as we talked about above.

Christian Kaltepoth <notifications@github.com> schrieb am Di., 10. Sep.
2019, 11:33:

**Comment by erdlet** [comment]
So it seems that the new approach works. The `Toolbox` receives a `PropertiesTranslationResolver` with the common `messages.properties` configuration and the showcase has an example for additional `TranslationResolver` like you mentioned in your example. I think that design is a really nice and flexible solution.

File: core/src/main/java/de/chkal/mvctoolbox/core/translation/DefaultTranslationResolverFactory.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+   * In case an application needs a more advanced setup or custom {@link TranslationResolver}s, an additional factory with
+   * producer methods can be created.
+   *
+   * @param mvcContext the {@link MvcContext} of the current request
+   * @return the default {@link PropertiesTranslationResolver} for the {@link de.chkal.mvctoolbox.core.Toolbox} bean
+   */
+  @Produces
+  @DefaultTranslationResolver
+  @ApplicationScoped
+  public PropertiesTranslationResolver getDefaultTranslationResolver(final MvcContext mvcContext) {
```
Comment (chkal) [inline core/src/main/java/de/chkal/mvctoolbox/core/translation/DefaultTranslationResolverFactory.java]: I'm not 100% sure, but AFAIK if we replace the return type with `TranslationResolver`, CDI will use the interface to create proxies and therefore we can remove the public default constructor of `PropertiesTranslationResolver`.

File: core/src/main/java/de/chkal/mvctoolbox/core/translation/PropertiesTranslationResolver.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.ResourceBundle;
+import javax.mvc.MvcContext;
+
+/**
+ * Resolves translations from a resource bundle with a configurable base name. The {@link Locale} for the
+ * processed request will be resolved by the applied {@link MvcContext}.
+ *
+ * @author Tobias Erdle
+ */
+public class PropertiesTranslationResolver implements TranslationResolver {
```
Comment (chkal) [inline core/src/main/java/de/chkal/mvctoolbox/core/translation/PropertiesTranslationResolver.java:15]: We could think about moving some base functionality to a base class `AbstractTranslationResolver` so other implementations can benefit from it. Especially the MessageFormat handling. But let's do this later after this one got merged.

File: core/src/main/java/de/chkal/mvctoolbox/core/translation/TranslationResolver.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package de.chkal.mvctoolbox.core.translation;
+
+import java.util.Locale;
+
+/**
+ * An interface for resolving translations based on a key and a {@link java.util.Locale}.
```
Comment (chkal) [inline core/src/main/java/de/chkal/mvctoolbox/core/translation/TranslationResolver.java:6]: Maybe add a note that it is the responsibility of the implementation to lookup the current locale?

File: showcase/pom.xml [hunk 4]
```diff
@@ -103,7 +112,7 @@
   </dependencies>
 
   <build>
-    <finalName>${project.artifactId}</finalName>
+    <finalName>showcase</finalName>
```
Comment (chkal) [inline showcase/pom.xml:115]: Good idea.

File: showcase/src/main/java/de/chkal/mvctoolbox/showcase/input/InputForm.java [hunk 5]
```diff
@@ -1,17 +1,17 @@
...
-import org.hibernate.validator.constraints.Length;
+import javax.validation.constraints.Size;
 
 import javax.ws.rs.FormParam;
 
 public class InputForm {
 
   @FormParam("text1")
-  @Length(min = 5, message = "Text #1: Please enter at least 5 characters")
+  @Size(min = 5, message = "Text #1: Please enter at least 5 characters")
```
Comment (chkal) [inline showcase/src/main/java/de/chkal/mvctoolbox/showcase/input/InputForm.java:10]: Good catch!

**Review by chkal — COMMENTED** [review]
Awesome work! Thanks a lot! This looks great. I just added minor comments. But I guess we are almost ready for merge!

File: core/src/main/java/de/chkal/mvctoolbox/core/translation/DefaultTranslationResolverFactory.java (hunk 1 above)
Comment (erdlet) [reply core/src/main/java/de/chkal/mvctoolbox/core/translation/DefaultTranslationResolverFactory.java]: Ah good catch, I think that was my problem the whole time.

File: core/src/main/java/de/chkal/mvctoolbox/core/translation/PropertiesTranslationResolver.java (hunk 2 above)
Comment (erdlet) [reply core/src/main/java/de/chkal/mvctoolbox/core/translation/PropertiesTranslationResolver.java:15]: :+1:

File: core/src/main/java/de/chkal/mvctoolbox/core/translation/TranslationResolver.java (hunk 3 above)
Comment (erdlet) [reply core/src/main/java/de/chkal/mvctoolbox/core/translation/TranslationResolver.java:6]: Yes good idea, I forgot that.

File: showcase/pom.xml (hunk 4 above)
Comment (erdlet) [reply showcase/pom.xml:115]: To be honest I did this only to be able to use Adam Biens `wad` script :sweat_smile: But thanks ;)

File: showcase/src/main/java/de/chkal/mvctoolbox/showcase/input/InputForm.java (hunk 5 above)
Comment (erdlet) [reply showcase/src/main/java/de/chkal/mvctoolbox/showcase/input/InputForm.java:10]: Thanks :+1:

File: core/src/main/java/de/chkal/mvctoolbox/core/translation/DefaultTranslationResolverFactory.java (hunk 1 above)
Comment (erdlet) [reply core/src/main/java/de/chkal/mvctoolbox/core/translation/DefaultTranslationResolverFactory.java]: Done

**Comment by chkal** [comment]
Awesome! Thanks a lot! Merging this now! Great work! :beers:

**Comment by chkal** [comment]
Maybe we should now add some documentation? Currently, the docs are in the wiki of this repository. Not sure if there is a way to prepare a pull request for the wiki to be honest.

**Pull request merged**